### DISCUSSION TITLE
LocalSearchFilter is not respected during First Solution phase

### DISCUSSION THREAD

**Discussion by glebarez** [google/or-tools#3389] [Routing (and legacy CP) questions]
Hello,
I noticed that my custom ```LocalSearchFilter``` is not used during the first solution phase, which leads to overall incorrect results.
The first solution sets the model variables to specific values, and if some of those values are not touched at later LocalSearch phase, I have no chance to deny the prohibited combinations from final outcome.
With above mechanics, the only reliable method for first solution is ```ALL_UNPERFORMED```, which doesn't yield results as good as other methods

**What is the proper way to apply filters during the first solution phase?**

**Comment by Mizux** [comment]
Is it possible to add some constraints instead ? using the underlaying solver `routing.solver().Add(....)` ?

**Reply by glebarez** [reply]
@Mizux Actually the constraint aproach didn't work in first place. (To restrict certain combinations I used AllowedAssignments, but solver delivered suboptimal solutions.) That's why I went implementing the filter, which actually works great, except for the FirstSolution phase..

**Reply by glebarez** [reply]
I will come back soon with more details

**Reply by glebarez** [reply]
@Mizux 
After investigating this further I can conclude that:
- FirstSolution decision builder actually consults the custom filter, but it doesn't pass the ```VehicleVar```s into it (which are the ones the filter is set to handle)

By the way I had the same problem during the local search phase, e.g. solver simply didn't optimize the ```VehicleVar```s because ```CostsAreHomogeneousAcrossVehicles()``` returned ```true``` (so it only optimized the ```nexts```)
https://github.com/google/or-tools/blob/49b6301e1e1e231d654d79b6032e79809868a70e/ortools/constraint_solver/routing.cc#L5553-L5569

I hacked this by setting this model parameter:
```python
 model_params.reduce_vehicle_cost_model = False
```
after this, ```VehicleVar```s were added to the solver, and delivered into LocalSearchFilter.
But this obviously doesn't work for FirstSolution decision builder.
PS: I am using ```PARALLEL_CHEAPEST_INSERTION```
